proposals/NNNN-library-evolution.md

```diff
@@ -0,0 +1,367 @@
+
+- it introduces indirection and function call overhead when accessing fields;
+- it means the fields of the struct must be stored on the heap even when they could otherwise be stored on the stack, with that heap allocation needing to be managed via reference counting or some other mechanism;
+- it means that the fields of the struct data are not stored in contiguous memory when placed in an array; and 
+- the implementation of all this must be provided by the library author.
+
+A similar challenge occurs with Swift enums. As discussed in [SE-0192][], introducing new cases to an enum can break source compatibility. There are also ABI consequences to this: adding new enum cases sometimes means increasing the storage needed for the enum, which can affect the size and layout of an enum in the same way adding fields to a struct can.
+
+The goal of this proposal is to reduce the burden on library developers by building into the compiler an automatic mechanism to reserve the flexibility to alter the internal representation of structs and enums, without manual workarounds. This mechanism can be implemented in such a way that optimizations such as stack allocation or contiguous inline storage of structs can still happen, while leaving the size of the type to be determined at runtime. 
+
+  [SE-0192]: https://github.com/apple/swift-evolution/blob/master/proposals/0192-non-exhaustive-enums.md
+
+## Proposed solution
+
+The compiler will gain a new mode, called "library evolution mode". This mode will be off by default. A library compiled with this mode enabled is said to be _ABI-stable_.
+
+When a library is compiled with library evolution mode enabled, it is not an ABI-breaking change to modify the fields in a struct (to add, remove, or reorder them), and to add new enum cases (including with associated values). This implies that clients must manipulate fields and enum cases indirectly, via non-inlinable function calls. Information such as the size of the type, and the layout of its fields, becomes something that can only be determined at runtime. Types that reserve this flexibility are referred to as "resilient types".
+
+A new `@frozen` keyword will be introduced to allow library authors to opt out of this flexibility on a per-type basis. This keyword promises that stored instance properties within the struct will not be *added,* *removed,* or *reordered*, and that an enum will never *add,* *remove,* or *reorder* its cases (note removing, and sometimes reordering, cases can already be source breaking, not just ABI breaking). The compiler will use this for optimization purposes when compiling clients of the type. The precise set of allowed changes is defined below.
```

Comment:
`@frozen` an attribute, not a keyword, correct?

Reply:
yes! thanks.

Reply:
> Okay, as an implicit contrast with the column to the left?

Yes. There are some things that affect the ABI even without frozen, which aren't bolded.

> I think it might be better to be stronger about this than "Affects".

Yeah, I guess "Not Allowed". The question would be, not allowed by whom, given the compiler isn't going to stop you. @jrose-apple do you have an opinion?

Reply:
(Somehow this ended up attached to the wrong comment)

"Not allowed" works for me.  "Allowed" has the saw actorship problem. :)